File: packages/core/types/task-view.ts
import { z } from "zod";

const JsonValue = z.unknown();

export const TaskViewSchema = z.object({
  id: z.string(),
  organization_id: z.string(),
  workspace_id: z.string(),
  owner_id: z.string(),
  name: z.string(),
  scope_type: z.string(),
  scope_id: z.string().nullable().optional(),
  scope_variant: z.string().nullable().optional(),
  visibility: z.string(),
  definition_version: z.number(),
  query: JsonValue,
  display: JsonValue,
  revision: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type TaskView = z.infer<typeof TaskViewSchema>;

export const TaskViewPreferenceSchema = z.object({
  scope_type: z.string(),
  scope_id: z.string(),
  prefs: JsonValue,
  updated_at: z.string().optional().default(""),
});
export type TaskViewPreference = z.infer<typeof TaskViewPreferenceSchema>;

export const TaskPinSchema = z.object({
  id: z.string(),
  organization_id: z.string(),
  workspace_id: z.string(),
  user_id: z.string(),
  item_type: z.string(),
  item_id: z.string(),
  position: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type TaskPin = z.infer<typeof TaskPinSchema>;

File: server/internal/handler/dto/sdo/task_view.go
package sdo

import "encoding/json"

// TaskViewDTO is one saved task view.
type TaskViewDTO struct {
	ID                string          `json:"id"`
	OrganizationID    string          `json:"organization_id"`
	WorkspaceID       string          `json:"workspace_id"`
	OwnerID           string          `json:"owner_id"`
	Name              string          `json:"name" example:"Open backlog"`
	ScopeType         string          `json:"scope_type" example:"workspace"`
	ScopeID           *string         `json:"scope_id"`
	ScopeVariant      *string         `json:"scope_variant"`
	Visibility        string          `json:"visibility" example:"private"`
	DefinitionVersion int32           `json:"definition_version"`
	Query             json.RawMessage `json:"query"`
	Display           json.RawMessage `json:"display"`
	Revision          int64           `json:"revision"`
	CreatedAt         string          `json:"created_at"`
	UpdatedAt         string          `json:"updated_at"`
}

// TaskViewListSDO is GET .../task-views.
type TaskViewListSDO struct {
	Views []TaskViewDTO `json:"views"`
	Total int           `json:"total"`
}

// TaskViewSDO wraps one view.
type TaskViewSDO struct {
	View TaskViewDTO `json:"view"`
}

// TaskViewPreferenceSDO is GET/PUT .../task-view-preferences.
type TaskViewPreferenceSDO struct {
	ScopeType string          `json:"scope_type"`
	ScopeID   string          `json:"scope_id"`
	Prefs     json.RawMessage `json:"prefs"`
	UpdatedAt string          `json:"updated_at,omitempty"`
}

// TaskPinDTO is one sidebar pin.
type TaskPinDTO struct {
	ID             string  `json:"id"`
	OrganizationID string  `json:"organization_id"`
	WorkspaceID    string  `json:"workspace_id"`
	UserID         string  `json:"user_id"`
	ItemType       string  `json:"item_type" example:"task"`
	ItemID         string  `json:"item_id"`
	Position       float64 `json:"position"`
	CreatedAt      string  `json:"created_at"`
	UpdatedAt      string  `json:"updated_at"`
}

// TaskPinListSDO is GET .../pins.
type TaskPinListSDO struct {
	Pins  []TaskPinDTO `json:"pins"`
	Total int          `json:"total"`
}

// TaskPinSDO wraps one pin.
type TaskPinSDO struct {
	Pin TaskPinDTO `json:"pin"`
}
